jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -1916,6 +1916,8 @@ public boolean release()
                     buffer.release();
                 _buffers.clear();
                 _aggregate = null;
+                // Re-acquire the ref counter to make sure this instance is reusable.
+                ((ReferenceCounter)getWrapped()).acquire();
```

Comment:
This is the actual fix here and the change to super above is not strictly necessary?

If this is needed here, then why is it not also needed for all other extensions of RBB.Abstract?

Reply:
Actually I really don't understand this fix?  If we have been released to 0, then we should not be reusable.   Maybe we can reacquire is a reset method?

Reply:
Maybe all the add methods of `RBB.DC` should have:
```java
            assert getWrapped() instanceof ReferenceCounter counter && counter.get() > 0;
```

or even a real check to that effect?


Ultimately `RBB.DC` is just a `RBB` and that it is a programming error to keep using it after releasing to zero.   If it was a pooled buffer, then you'd be using a buffer that might be used by a different process.      If it is being used as an accumulator that will be reused over and over, but release by a method it is passed to, then you need to retain before you pass it.

Reply:
The change to the constructor calling super with a `ReferenceCounter` instance is to guarantee the cast in `release()` won't fail.

The test that highlighted this problem does the following:

```
RBB.DC.put(...)
RBB.DC.writeTo(...)
RBB.DC.release(...)

RBB.DC.put(...)
RBB.DC.writeTo(...)
RBB.DC.release(...)
```

You're totally right that using a RBB after the final `release()` is an error, and we should consider adding some checks to our implementations to prevent the usage of any RBB impl after release.

But in this particular case, I noticed that altering `release()` to make it reset the wrapped `ReferenceCounter` was all it took for `RBB.DC` to behave correctly, so I figured that since it was be simpler to make `RBB.DC` behave correctly despite being incorrectly used then preventing its incorrect use I'd go with that solution.

But I agree this is debatable, and that opens some questions: should we modify the `RBB` contract to state it throws when used after release returned true? Is there anything wrong with allowing certain implementations to allow being reused after release returned true?

Reply:
I think there is something wrong with that approach. If code is written without knowledge of the specific type of RBB, then it will work with one implementation but fail with another

Reply:
Sure, and since this is happening in the context of a test, we are free to decide to either modify the test to strictly follow the RBB contract or to leave the test the way it is if we consider this RBB.DC usage is reasonable.

Reply:
Let's do that and add the asserts I suggested